misk/src/main/kotlin/misk/CoordinatedService2.kt

```diff
@@ -0,0 +1,238 @@
+ * Symmetrically it stalls in the `STOPPING` state until all dependent services are `TERMINATED`.
+ */
+internal class CoordinatedService2(val service: Service) : AbstractService() {
+  /** Services that start before this. */
+  private val directDependsOn = mutableSetOf<CoordinatedService2>()
+
+  /** Services this starts before. */
+  private val directDependencies = mutableSetOf<CoordinatedService2>()
+
+  /**
+   * Services that enhance this. This starts before them, but they start before
+   * [directDependencies].
+   */
+  private val enhancements = mutableSetOf<CoordinatedService2>()
+
+  /** Service that starts up before this, and whose [directDependencies] also depend on this. */
+  private var enhancementTarget: CoordinatedService2? = null
+
+  init {
+    service.addListener(object : Listener() {
+      override fun running() {
+        synchronized(this) {
+          notifyStarted()
+        }
+        downstreamServices.forEach { it.startIfReady() }
+      }
+
+      override fun terminated(from: State) {
+        synchronized(this) {
+          notifyStopped()
+        }
+        upstreamServices.forEach { it.stopIfReady() }
+      }
+
+      override fun failed(from: State, failure: Throwable) {
+        notifyFailed(failure)
+      }
+    }, MoreExecutors.directExecutor())
+  }
+
+  /**
+   * Returns a set of services that are required by this service.
+   *
+   * The set consists of the [enhancementTarget], all direct dependencies and each dependency's
+   * transitive enhancements. It is the set of services that block start-up of this service.
+   */
+  val upstreamServices: Set<CoordinatedService2> by lazy {
+    val result = mutableSetOf<CoordinatedService2>()
+    if (enhancementTarget != null) {
+      result += enhancementTarget!!
+    }
+    for (provider in directDependsOn) {
+      result += provider
+      provider.getTransitiveEnhancements(result)
+    }
+    result
+  }
+
+  private fun getTransitiveEnhancements(sink: MutableSet<CoordinatedService2>) {
+    sink += enhancements
+    for (enhancement in enhancements) {
+      enhancement.getTransitiveEnhancements(sink)
+    }
+  }
+
+  /**
+   * Returns a set of all services which require this service to be started before they can start,
+   * and who must shut down before this service shuts down.
+   *
+   * The set contains this service's enhancements, its dependencies, and the dependencies of
+   * [enhancementTarget] (if it exists) and all transitive targets.
+   */
+  val downstreamServices: Set<CoordinatedService2> by lazy {
+    val result = mutableSetOf<CoordinatedService2>()
+    result += enhancements
+    var t: CoordinatedService2? = this
+    while (t != null) {
+      result += t.directDependencies
+      t = t.enhancementTarget
+    }
+    result
+  }
+
+  /** Adds [services] as dependents downstream. */
+  fun addDependencies(vararg services: CoordinatedService2) {
+    directDependencies += services
```

Comment:
This seems dangerous. I don't think we want to allow dynamically added dependencies at runtime - dependencies should be built during graph initialization and the made immutable prior to service start. Ideally the service graph construction would occur outside the service itself such that by the time the service is created it is immutable; failing that we should be checking invariants here and failing if you attempt to add a dependency after the service is started.

Reply:
@mmihic due to the self-relationships in the class, it doesn't seem possible to add enhancements/dependencies at construction time, but commit 10e38a8 adds runtime checks (and tests) that:
 - no _`CoordinatedService2`_ can wrap an already running Service
 - no _`CoordinatedService2`_ can have a dependency/enhancement added if it or the services to be added to it are not all `NEW`

Hopefully this addresses your concerns!